File: Multi-Threading-Python/bubbleSortFunction.py
def bubble_sort(input_array, descending=False):
    """
    Sort an array using the Bubble Sort algorithm.

    Parameters:
        input_array (list): The array to sort.
        descending (bool): If True, sort the array in descending order. Default is False.

    Returns:
        list: The sorted array.

    Raises:
        ValueError: If input_array is not a list.
    """

    if not isinstance(input_array, list):
        raise ValueError("Input must be a list")

    array = input_array.copy()  
    n = len(array)

    for i in range(n):
        swapped = False
        lastindexswapped = n - 1

        for j in range(0, lastindexswapped):
            if (array[j] < array[j + 1]) if descending else (array[j] > array[j + 1]):
                array[j], array[j + 1] = array[j + 1], array[j]
                swapped = True
                lastindexswapped = j

        if not swapped: 
            break

    return array

File: javaProject/BubbleSortFunction.java
import java.util.ArrayList;
import java.util.List;

public class BubbleSortFunction {
	
	/**
	 * Sort an array using the Bubble Sort algorithm
	 * 
	 * @param List<Integer> to sort
	 * @return List<Integer> sorted array
	 */
	public List<Integer> bubbleSort(List<Integer> inputArray) {
		// Input validation
		if (inputArray == null || inputArray.size() < 0) {
			throw new IllegalArgumentException("Invalid input array");
		}
		
		boolean swapped;
		
		// Create a copy of the input array to avoid modifying it
		List<Integer> array = new ArrayList<>(inputArray);
		
		for (int i = 0; i < array.size(); i++) {
			
			// Boolean variable is initialized to false so that it is then checked if there was another swap
			swapped = false;
			
			for (int j = 0; j < array.size() - 1; j++) {
				if (array.get(j) > array.get(j+1)) {
					// Swap array[j] and array[j+1]
					int aux = array.get(j);
					array.set(j, array.get(j+1));
					array.set(j+1, aux);
					swapped = true;
				}
			}
			
			if(!swapped) {
				break;
			}
		}
		
		return array;
	}
}
